src/simd/definitions.h

```diff
@@ -0,0 +1,124 @@
+/*
+ * Copyright 2017-2018 Scality
+ *
+ * Redistribution and use in source and binary forms, with or without
+ * modification, are permitted provided that the following conditions are met:
+ *
+ * 1. Redistributions of source code must retain the above copyright notice,
+ *    this list of conditions and the following disclaimer.
+ *
+ * 2. Redistributions in binary form must reproduce the above copyright notice,
+ *    this list of conditions and the following disclaimer in the documentation
+ *    and/or other materials provided with the distribution.
+ *
+ * 3. Neither the name of the copyright holder nor the names of its contributors
+ *    may be used to endorse or promote products derived from this software
+ *    without specific prior written permission.
+ *
+ * THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS IS"
+ * AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
+ * IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE
+ * ARE DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT HOLDER OR CONTRIBUTORS BE
+ * LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR
+ * CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF
+ * SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS
+ * INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN
+ * CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE)
+ * ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE
+ * POSSIBILITY OF SUCH DAMAGE.
+ */
+
+/** @file definitions.h
+ *
+ * Provide includes and constants that are platform-specific.
+ */
+
+#ifndef __QUAD_SIMD_SIMD_DEFINITIONS_H__
+#define __QUAD_SIMD_SIMD_DEFINITIONS_H__
+
+#include <climits>
+#include <cstdint>
+
+// SIMD-specific includes {{{
+
+#ifdef QUADIRON_USE_SIMD
+
+// SIMD for x86 and x86_64
+#if defined(__i386__) || defined(__x86_64__)
+#include <x86intrin.h>
+#endif
+
+// SIMD for ARM
+#if defined(__ARM_NEON)
+#include <arm_neon.h>
+#endif
+
+#endif
+
+// }}}
+
+namespace quadiron {
+namespace simd {
+
+/// Supported instruction set.
+enum class InstructionSet {
+    NONE, ///< No SIMD instruction (fallback).
+    SSE,  ///< SSE4.1
+    AVX,  ///< AVX2
+};
+
+// Definitions for Intel AVX-256 {{{
+
+// We required AVX2 because we relies on some instructions (such as
+// `_mm256_add_epi16` and others) that aren't available in the first version of
+// AVX.
+#if defined(__AVX__) && defined(__AVX2__)
+
+using RegisterType = __m256;
+using MaskType = __m256i;
+
+static constexpr InstructionSet INSTRUCTION_SET = InstructionSet::AVX;
+
+// }}}
+// Definitions for Intel SSE {{{
+
+#elif defined(__SSE__) && defined(__SSE4_1__)
+
+using RegisterType = __m128;
+using MaskType = __m128i;
+
+static constexpr InstructionSet INSTRUCTION_SET = InstructionSet::SSE;
+
+// }}}
+// Definitions for scalar fallback {{{
```

Comment:
Do we need definitions for non-supported SIMD?

Reply:
Yeah, because my idea to reduce code duplication is to try, as much as possible, to have a single code path.

Currently we have many code duplication with:
- `f1` that implement function without SIMD
- `f2.1` which implement function with SSE on 16-bit
- `f2.2` which implement function with SSE on 32-bit
- `f3.1` which implement function with AVX on 16-bit
- `f3.2` which implement function with AVX on 32-bit

So basically, 5 functions (and we don't support yet NEON and AVX256, nor speedup on 64-bit) that are doing the same thing.

If we could have a generic `Register` class with the necessary methods we could write code that could look like

```c++
for (unsigned i = 0; i < data.size(); i += simd::countof<T>()) {
    Register<T> r1(&data[i]);
    Register<T> r2(&coef[i]);

    Register<T> out(&output[i]);
    out = r1 + r2;
}
```

And then, if it's compiled with SIMD the constructor would call the correct ` _mm*_load*` instrisic, the `operator+` would be implemented by using the correct `_mm*_add*` instrinsics, …
And if we compile without SIMD, then `Register` would simply iterate element by element, using native multiplication, …

That way, we could write a single function and still benefit from SIMD acceleration.
All the specific would be hidden in the `Register` class (using template specialization) and our "main" code (the FEC) would be more readable.

I don't know yet how the final code will look like, I'm still working on the implementation, but that's the goad of this whole "rework the SIMD API".

Reply:
Yes, it looks good. It looks similar what https://github.com/QuantStack/xsimd does
BTW, in my PR #254, all 5 functions are combined in a templated one. There are still only wrappers of SIMD operations and basic operations that depend on register type and 16/32-bit packed elements

Reply:
I don't know this lib specifically, but yeah I think it's a pretty standard approach.

> BTW, in my PR #254, all 5 functions are combined in a templated one.

On one hand, it's good to hear, on the other hand I'm not sure it's a good idea to have two concurrent PR reworking the SIMD :S
I hope we won't conflict too much…

Reply:
I think there are only small conflicts. Your PR is on SIMD definitions. In PR #254, I re-implement all modular operations and vectorized functions for Radix2 FFT. So I re-organize these codes to remove duplicated functions.